### PR TITLE
WIP: Add support for files generated by appstreamcli compose

### PR DESCRIPTION
**Pull request by barthalion**

### CODE REVIEW

**Comment by ximion** [comment]
Looks good to me!

**Comment by TingPing** [comment]
@barthalion Is this still WIP?

File: app/flatpak-builtins-build-bundle.c [hunk 1]
```diff
@@ -127,9 +127,13 @@ get_bundle_appstream_data (GFile        *root,
 
   *result = NULL;
 
-  xmls_dir = g_file_resolve_relative_path (root, "files/share/app-info/xmls");
-  appstream_basename = g_strconcat (name, ".xml.gz", NULL);
-  appstream_file = g_file_get_child (xmls_dir, appstream_basename);
+  appstream_file = g_file_resolve_relative_path (root, "files/share/swcatalog/xml/flatpak.xml.gz");
+  if (!g_file_test (g_file_get_path (appstream_file), G_FILE_TEST_EXISTS))
```
Comment (TingPing) [inline app/flatpak-builtins-build-bundle.c:131]:
```suggestion
  if (!g_file_test (g_file_peek_path (appstream_file), G_FILE_TEST_EXISTS))
```

Fixes leak

File: app/flatpak-builtins-build-bundle.c [hunk 2]
```diff
@@ -127,9 +127,13 @@ get_bundle_appstream_data (GFile        *root,
...
+    {
+      xmls_dir = g_file_resolve_relative_path (root, "files/share/app-info/xmls");
+      appstream_basename = g_strconcat (name, ".xml.gz", NULL);
+      appstream_file = g_file_get_child (xmls_dir, appstream_basename);
+    }
```
Comment (TingPing) [inline app/flatpak-builtins-build-bundle.c:136]:
```suggestion
    {
      g_clear_object (&appstream_file);
      xmls_dir = g_file_resolve_relative_path (root, "files/share/app-info/xmls");
      appstream_basename = g_strconcat (name, ".xml.gz", NULL);
      appstream_file = g_file_get_child (xmls_dir, appstream_basename);
    }
```

Fixes leak.

File: common/flatpak-dir.c [hunk 3]
```diff
@@ -3365,8 +3365,12 @@ read_appdata_xml_from_deploy_dir (GFile *deploy_dir, const char *id)
...
   gsize size;
 
-  appdata_name = g_strconcat (id, ".xml.gz", NULL);
-  appdata_file  = flatpak_build_file (deploy_dir, "files/share/app-info/xmls", appdata_name, NULL);
+  appdata_file = flatpak_build_file (deploy_dir, "files/share/swcatalog/xml/flatpak.xml.gz",  NULL);
+  if (!g_file_test (g_file_get_path (appdata_file), G_FILE_TEST_EXISTS))
+    {
+      appdata_name = g_strconcat (id, ".xml.gz", NULL);
+      appdata_file = flatpak_build_file (deploy_dir, "files/share/app-info/xmls", appdata_name, NULL);
+    }
```
Comment (TingPing) [inline common/flatpak-dir.c:3373]:
```suggestion
  appdata_file = flatpak_build_file (deploy_dir, "files/share/swcatalog/xml/flatpak.xml.gz",  NULL);
  if (!g_file_test (g_file_peek_path (appdata_file), G_FILE_TEST_EXISTS))
    {
      g_clear_object (&appstream_file);
      appdata_name = g_strconcat (id, ".xml.gz", NULL);
      appdata_file = flatpak_build_file (deploy_dir, "files/share/app-info/xmls", appdata_name, NULL);
    }
```

File: common/flatpak-installed-ref.c [hunk 4]
```diff
@@ -535,7 +537,15 @@ flatpak_installed_ref_load_appdata (FlatpakInstalledRef *self,
...
 
   appdata_name = g_strconcat (flatpak_ref_get_name (FLATPAK_REF (self)), ".xml.gz", NULL);
-  path = g_build_filename (priv->deploy_dir, "files/share/app-info/xmls", appdata_name, NULL);
+  appinfo_path = g_build_filename (priv->deploy_dir, "files/share/app-info/xmls", appdata_name, NULL);
+  swcatalog_path = g_build_filename (priv->deploy_dir, "files/share/swcatalog/xml/flatpak.xml.gz", NULL);
+
+  if (g_file_test (appinfo_path, G_FILE_TEST_EXISTS))
+    path = appinfo_path;
+  else if (g_file_test (swcatalog_path, G_FILE_TEST_EXISTS))
+    path = swcatalog_path;
```
Comment (TingPing) [inline common/flatpak-installed-ref.c:546]:
```suggestion
  if (g_file_test (appinfo_path, G_FILE_TEST_EXISTS))
    path = g_steal_pointer (&appinfo_path);
  else if (g_file_test (swcatalog_path, G_FILE_TEST_EXISTS))
    path = g_steal_pointer (&swcatalog_path);
```

Fixes double free.

File: common/flatpak-utils.c [hunk 5]
```diff
@@ -5366,14 +5366,19 @@ extract_appstream (OstreeRepo        *repo,
...
+  appstream_file = g_file_resolve_relative_path (root, "files/share/swcatalog/xml/flatpak.xml.gz");
+  if (g_file_test (g_file_get_path (appstream_file), G_FILE_TEST_EXISTS))
+    app_info_dir = g_file_resolve_relative_path (root, "files/share/swcatalog");
+  else
+    {
+      app_info_dir = g_file_resolve_relative_path (root, "files/share/app-info");
+      xmls_dir = g_file_resolve_relative_path (app_info_dir, "xmls");
+      appstream_basename = g_strconcat (id, ".xml.gz", NULL);
+      appstream_file = g_file_get_child (xmls_dir, appstream_basename);
+    }
```
Comment (TingPing) [inline common/flatpak-utils.c:5378]:
```suggestion
  if (g_file_test (g_file_peek_path (appstream_file), G_FILE_TEST_EXISTS))
    app_info_dir = g_file_resolve_relative_path (root, "files/share/swcatalog");
  else
    {
      g_clear_object (&appstream_file);
      app_info_dir = g_file_resolve_relative_path (root, "files/share/app-info");
      xmls_dir = g_file_resolve_relative_path (app_info_dir, "xmls");
      appstream_basename = g_strconcat (id, ".xml.gz", NULL);
      appstream_file = g_file_get_child (xmls_dir, appstream_basename);
    }
```

File: app/flatpak-builtins-build-bundle.c [hunk 6]
```diff
@@ -127,9 +127,13 @@ get_bundle_appstream_data (GFile        *root,
 
   *result = NULL;
 
-  xmls_dir = g_file_resolve_relative_path (root, "files/share/app-info/xmls");
-  appstream_basename = g_strconcat (name, ".xml.gz", NULL);
-  appstream_file = g_file_get_child (xmls_dir, appstream_basename);
+  appstream_file = g_file_resolve_relative_path (root, "files/share/swcatalog/xml/flatpak.xml.gz");
```
Comment (TingPing) [inline app/flatpak-builtins-build-bundle.c:130]: You repeat this chunk of code 3 times. I think it could be pulled out into a function.

**Comment by barthalion** [comment]
I still consider it a draft until I manage to test it with flatpak-builder's counterpart PR and double-check resulting builds/repo are the same.

**Comment by TingPing** [comment]
I'll close this as superseded by #5277

**Pull request closed without merging**